luda-adventure/src/app/game/update/calculate_next_motion_of_character/calculate_next_movement_of_character.rs

```diff
@@ -0,0 +1,279 @@
+use crate::{
+    app::game::{
+        known_id::object::PLAYER_CHARACTER, simplify_collision_box_list, zero_speed, Collider,
+        CollisionBox, MovementPath, PlayerCharacter, Positioner, TileExt, Velocity,
+    },
+    ecs,
+};
+use float_cmp::{ApproxEq, F32Margin};
+use namui::prelude::*;
+
+const COLLISION_MARGIN: F32Margin = F32Margin {
+    ulps: 1,
+    epsilon: 0.0001,
+};
+
+pub fn calculate_next_movement_of_character(ecs_app: &mut ecs::App, max_duration: Time) {
+    let target_collider_list = ecs_app.query_entities::<(&Collider, &Positioner)>();
+    let target_collision_box_list = simplify_collision_box_list(
+        target_collider_list
+            .iter()
+            .filter_map(|(target_entity, (target_collider, target_positioner))| {
+                if target_entity.id() == PLAYER_CHARACTER {
+                    None
+                } else {
+                    Some(target_collider.get_collision_box(target_positioner.xy(0.ms())))
```

Comment:
Why it use 0.ms(), not the current time to calculate? It's kind of TODO?

Reply:
Yes, due to rust's borrow strategy.
Outdated